lib/experimental/block-hooks.php

```diff
@@ -158,14 +158,28 @@ function gutenberg_insert_hooked_block( $inserted_block, $relative_position, $an
 				array_unshift( $block['innerBlocks'], $inserted_block );
 				// Since WP_Block::render() iterates over `inner_content` (rather than `inner_blocks`)
 				// when rendering blocks, we also need to prepend a value (`null`, to mark a block
-				// location) to that array.
-				array_unshift( $block['innerContent'], null );
+				// location) to that array after HTML content for the inner blocks wrapper.
+				$chunk_index = 0;
+				for ( $index = $chunk_index; $index < count( $block['innerContent'] ); $index++ ) {
+					if ( is_null( $block['innerContent'][ $index ] ) ) {
+						$chunk_index = $index;
+						break;
+					}
+				}
+				array_splice( $block['innerContent'], $chunk_index, 0, array( null ) );
 			} elseif ( 'last_child' === $relative_position ) {
 				array_push( $block['innerBlocks'], $inserted_block );
 				// Since WP_Block::render() iterates over `inner_content` (rather than `inner_blocks`)
-				// when rendering blocks, we also need to prepend a value (`null`, to mark a block
-				// location) to that array.
-				array_push( $block['innerContent'], null );
+				// when rendering blocks, we also need to correctly prepend a value (`null`, to mark a block
+				// location) to that array after HTML content for the inner blocks wrapper.
+				$chunk_index = count( $block['innerContent'] ) - 1;
+				for ( $index = $chunk_index; $index >= 0; $index-- ) {
+					if ( is_null( $block['innerContent'][ $index ] ) ) {
+						$chunk_index = $index;
+						break;
+					}
+				}
```

Comment:
Noting that this sets `$chunk_index` to `-1` if `$block['innerContent']` is empty. Which I think is okay, since `array_splice()` also works if the offset arg is negative ([see](https://www.php.net/manual/en/function.array-splice.php)).

Reply:
Good catch. I think I should set `$chunk_index` to 0 initially and only update when `null` is in the array. I now realized that it's possible that the array is empty 😄 

I was studying unit tests for the parser earlier today, some interesting findings for `innerContent`:

https://github.com/WordPress/gutenberg/blob/7652de9ed159d4e9698e432c06b83e56777160c8/packages/block-serialization-spec-parser/shared-tests.js#L241-L289

In particular, the fact that inner block locations always become `null`, so you can have multiple of them. However, if there are no inner blocks then we won't have `null`, and it can even be an empty array 🤔 

I need to study some existing blocks to see how they serialize when there are no inner blocks. Group, Columns, Column, it might get tricky. I'm wondering if at least for core blocks we always enforce a HTML wrapper (I would assume yes, but it isn't a requirement for the parser), and whether we show the wrapper when there are no inner blocks inserter. I hope that with Columns that can be explored.

Reply:
An example for the Columns block with 3 columns and the middle one has no child blocks:

```html
<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>First</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Third</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
```

A Group block with no child blocks:

```html
<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"></div>
<!-- /wp:group -->
```